src/mlpack/methods/ann/layer/lstm_impl.hpp

```diff
@@ -31,351 +26,425 @@ LSTM<InputDataType, OutputDataType>::LSTM()
+      (1 + arma::exp(-inputGate.cols(forwardStep, forwardStep + batchStep)));
+
+  forgetGateActivation.cols(forwardStep, forwardStep + batchStep) = 1.0 /
+      (1 + arma::exp(-forgetGate.cols(forwardStep, forwardStep + batchStep)));
+
+  hiddenLayer.cols(forwardStep, forwardStep + batchStep) = input2HiddenWeight *
+      input + output2HiddenWeight * outParameter.cols(
+      forwardStep, forwardStep + batchStep);
+
+  hiddenLayer.cols(forwardStep, forwardStep + batchStep).each_col() +=
+      input2HiddenBias;
+
+  hiddenLayerActivation.cols(forwardStep, forwardStep + batchStep) =
+      arma::tanh(hiddenLayer.cols(forwardStep, forwardStep + batchStep));
+
+  if (forwardStep == 0)
   {
-    outParameter.push_back(output);
-    cellParameter.push_back(std::move(tempPrevCell));
-    prevOutput = --outParameter.end();
-    prevCell = --cellParameter.end();
+    cell.cols(forwardStep, forwardStep + batchStep) =
+        inputGateActivation.cols(forwardStep, forwardStep + batchStep) %
+        hiddenLayerActivation.cols(forwardStep, forwardStep + batchStep);
   }
   else
   {
-    if (forwardStep == 1)
-    {
-      outParameter.clear();
-      cellParameter.clear();
+    cell.cols(forwardStep, forwardStep + batchStep) =
+        forgetGateActivation.cols(forwardStep, forwardStep + batchStep) %
+        cell.cols(forwardStep - batchSize, forwardStep - batchSize + batchStep)
+        + inputGateActivation.cols(forwardStep, forwardStep + batchStep) %
+        hiddenLayerActivation.cols(forwardStep, forwardStep + batchStep);
+  }
 
-      outParameter.push_back(output);
-      cellParameter.push_back(std::move(tempPrevCell));
+  outputGate.cols(forwardStep, forwardStep + batchStep) = input2GateOutputWeight
+      * input + output2GateOutputWeight * outParameter.cols(
+      forwardStep, forwardStep + batchStep) + cell.cols(forwardStep,
+      forwardStep + batchStep).each_col() % cell2GateOutputWeight;
 
-      prevOutput = outParameter.begin();
-      prevCell = cellParameter.begin();
-    }
-    else
-    {
-      *prevOutput = output;
-      *prevCell = std::move(tempPrevCell);
-    }
+  outputGate.cols(forwardStep, forwardStep + batchStep).each_col() +=
+      input2GateOutputBias;
+
+  outputGateActivation.cols(forwardStep, forwardStep + batchStep) = 1.0 /
+      (1 + arma::exp(-outputGate.cols(forwardStep, forwardStep + batchStep)));
+
+  cellActivation.cols(forwardStep, forwardStep + batchStep) =
+      arma::tanh(cell.cols(forwardStep, forwardStep + batchStep));
+
+  outParameter.cols(forwardStep + batchSize,
+      forwardStep + batchSize + batchStep) =
+      cellActivation.cols(forwardStep, forwardStep + batchStep) %
+      outputGateActivation.cols(forwardStep, forwardStep + batchStep);
+
+  output = OutputType(outParameter.memptr() +
+      (forwardStep + batchSize) * outSize, outSize, batchSize, false, false);
+
+  forwardStep += batchSize;
+  if ((forwardStep / batchSize) == rho)
```

Comment:
Should this be `rhoSize` instead of `rho`?  When I was doing some timing simulations with this implementation, I had problems because `rho` was set to `SIZE_MAX` but `rhoSize` was 20, so `forwardStep` would never be reset.  It's possible there is something else I misunderstood.

Reply:
Nice catch, but since ``rho`` controls the maximum number of steps to backpropagate through time, it could be < ``rhoSize``, so we should use ``std::min(rho, rhoSize)``.